doc/source/data/training_ingest.rst

```diff
@@ -0,0 +1,525 @@
+    )
+    my_trainer.fit()
+
+.. testoutput::
+    :hide:
+
+    ... 
+
+The subclass must be serializable, since it will be copied from the driver script to the driving actor of the Trainer. Its :meth:`configure <ray.train.DataConfig.configure>` method is called on the main actor of the Trainer group to create the data iterators for each worker.
+
+In general, you can use :class:`DataConfig <ray.train.DataConfig>` for any shared setup that has to occur ahead of time before the workers start iterating over data. The setup will be run at the start of each Trainer run.
+
+Performance tips
+~~~~~~~~~~~~~~~~
+
+.. _dataset_cache_performance:
+
+Caching the preprocessed Dataset
+################################
+If you are training on GPUs and have an expensive CPU preprocessing operation, this may bottleneck training throughput.
+
+If your preprocessed Dataset is small enough to fit in object store memory, the easiest thing to do is to *materialize* the preprocessed dataset in Ray object store memory, by calling :meth:`materialize() <ray.data.Dataset.materialize>` on the preprocessed dataset. This tells Ray Data to compute the entire preprocessed and pin it in the Ray object store memory. As a result, when iterating over the dataset repeatedly, the preprocessing operations do not need to be re-run. However, the trade-off is that if the preprocessed data is too large to fit into Ray object store memory, this will greatly decrease performance as data needs to be spilled to disk.
+
+.. testcode::
+
+    # Load the data.
+    train_ds = ray.data.read_parquet("s3://anonymous@ray-example-data/iris.parquet")
+
+    # Preprocess the data. Transformations that are made before the materialize call
+    # below will only be run once.
+    train_ds = train_ds.map_batches(normalize_length)
+
+    # Materialize the dataset in object store memory.
+    # Only do this if train_ds is small enough to fit in object store memory.
+    train_ds = train_ds.materialize()
+
+    # Add per-epoch preprocessing. Transformations that you want to run per-epoch, such
+    # as data augmentation, should go after the materialize call.
+    train_ds = train_ds.map_batches(augment_data)
+
+    # Pass train_ds to the Trainer as normal...
+
+Adding CPU-only nodes to your cluster
+#####################################
+If you are bottlenecked on expensive CPU preprocessing and the preprocessed Dataset is too large to fit in object store memory, then the above tip will not work.
+
+In this case, since Ray supports heterogeneous clusters, you can add more CPU-only nodes to your cluster. This adds more CPU cores to your cluster to further parallelize preprocessing. It also adds more object store memory, allowing more data to be buffered in between preprocessing and training if training is bottlenecked by memory.
+
+If your preprocessing is more compute heavy than memory heavy, then add enough CPU only nodes to :ref:`cache the preprocessed dataset <dataset_cache_performance>`.
```

Comment:
```suggestion
In this case, since Ray supports heterogeneous clusters, you can add more CPU-only nodes to your cluster. This adds more CPU cores to your cluster to further parallelize preprocessing.

This also adds more object store memory that can be :ref:`used as a cache <dataset_cache_performance>`. This can be useful if your preprocessing is more compute-heavy than memory-heavy, and the preprocessed dataset doesn't fit in the memory of the GPU machines alone.
```

Reply:
Ah I wanted to make this distinction- for training pipelines that are bound by memory in between preprocessing and training, adding more CPU nodes will add more object store memory to buffer more preprocessed data. This does not necessarily need to fit the _entire_ preprocessed dataset, but just enough to not be bottlenecked by the buffering.

However, for training pipelines that are bottlenecked by compute, it would be better to just materialize the whole preprocessed dataset.

Reply:
Let me rewrite this to clarify more